### PR TITLE
[FEATURE] SCC and CCD encoder

### LINKED ISSUE #1120: [PROPOSAL] Add an option to output Scenarist Closed Caption format (*.SCC) and/or McPoodle's Disassembly format (*.CCD)
Please prefix your issue with one of the following: [BUG], [PROPOSAL], [QUESTION].

CCExtractor version (using the --version parameter preferably) : THIS WOULD APPLY TO VERSIONS AFTER **0.88**

**In raising this issue, I confirm the following (please check boxes, eg [X] - and delete unchecked ones):**

- [X] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [X] I have checked that the issue I'm posting isn't already reported.
- [X] I have checked that the issue I'm porting isn't already solved and no duplicates exist in [closed issues](https://github.com/CCExtractor/ccextractor/issues?q=is%3Aissue+is%3Aclosed) and in [opened issues](https://github.com/CCExtractor/ccextractor/issues)
- [X] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.
- [X] I have used the latest available version of CCExtractor to verify this issue exists.

**My familiarity with the project is as follows (check one, eg [X] - and delete unchecked ones):**

- [ ] I have never used CCExtractor.
- [ ] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [x] I am an active contributor to CCExtractor. (I have posted once before. I use CCExtractor A LOT.)

**Necessary information**
- Is this a regression (did it work before)? [X] NO | [ ] YES - *please specify the last known working version*
- What platform did you use? [X] Windows - [ ] Linux - [ ] Mac
- What were the used arguments? `-autoprogram` (not applicable yet)

**Video links (replace text below with your links)**

Please make the affected input file available for us (no screenshots, those don't help!). Public links to Dropbox, Google Drive, etc, are all fine. If it is not possible to make it available publicly, send us a private invitation (both Dropbox and Google Drive allow that). In this case we will download the file and upload it to the private developer repository.

Do *not* upload your file to any location that will require us to sign up or endure a wait list, slow downloads, etc. If your upload expires make sure you keep it active somehow (replace links if needed). Keep in mind that while we go over all tickets some may take a few days, and it's important we have the file available when we actually need it.

**Additional information**

In my spare time I've been writing a script (JavaScript in a web page for now) to convert closed caption disassembly files to advanced substation alpha.  It's a long story, but the punchline is I like to watch my shows with the captions looking like captions.  That means black background behind the text, monospaced fonts, coordinate-positioning, etc., but not all players/playback devices support all the features.

Usually I output into McPoodle's RAW format and command-line programs from SCC_TOOLS to convert to disassembly, but if there was a problem with reception when I recorded the show, then synchronization issues occur (RAW data being what it is). CCExtractor gives me no synchronization issues that I can see when outputting to text formats.

Anyway, my script is over 99% finished for pop-on captions, 0% for paint-on captions, and 0% for roll-up captions. I can post my work so far, if anyone's interested.

Thank you all for your time, and thank you for such a useful project that I use nearly every day.

PS: Make sure you set an alert in GitHub so you get notifications about your ticket. We may need to ask questions and we do everything inside GitHub's system.

### PR DESCRIPTION
**Pull request by NilsIrl**
Fix #1120 

- [X] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that another pull request for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.
- [x] **I have mentioned this change in the [changelog](https://github.com/CCExtractor/ccextractor/blob/master/docs/CHANGES.TXT).**

**My familiarity with the project is as follows (check one):**

- [X] I have used CCExtractor just a couple of times.

---

Colours haven't been implemented yet.

~~The disassembly format hasn't been implemented yet.~~

### CODE REVIEW

**Comment by NilsIrl** [comment]
The bitmap OCR hasn't been implemented either

**Comment by NilsIrl** [comment]
Dealing with different line endings hasn't been implemented yet either ("\n" vs "\r\n")

**Comment by NilsIrl** [comment]
Modifying subtitles hasn't been implemented yet. As mentioned in https://github.com/CCExtractor/ccextractor/pull/1139#issuecomment-566759227 it's probably better if it was done in a different way

**Comment by NilsIrl** [comment]
I will need more examples to be able to continue with the rest

File: src/lib_ccx/ccx_encoders_srt.c [hunk 1]
```diff
@@ -175,58 +175,57 @@ int write_cc_subtitle_as_srt(struct cc_subtitle *sub,struct encoder_ctx *context
...
-	millis_to_time (ms_start,&h1,&m1,&s1,&ms1);
-	millis_to_time (ms_end-1,&h2,&m2,&s2,&ms2); // -1 To prevent overlapping with next line.
+	millis_to_time(ms_start, &h1, &m1, &s1, &ms1);
+	millis_to_time(data->end_time - 1, &h2, &m2, &s2, &ms2); // -1 To prevent overlapping with next line.
 	char timeline[128];
-	context->srt_counter++;
-	sprintf(timeline, "%u%s", context->srt_counter, context->encoded_crlf);
-	used = encode_line(context, context->buffer,(unsigned char *) timeline);
+
+	sprintf(timeline, "%u%s", ++context->srt_counter, context->encoded_crlf);
```
Comment (cfsmp3) [inline src/lib_ccx/ccx_encoders_srt.c]: I don't love this change. Of course it's correct but it's less readable. An explicit change of the variable in its own line is clearer.

File: src/lib_ccx/ccx_common_common.c [hunk 2]
```diff
@@ -49,13 +49,13 @@ void millis_to_time(LLONG milli, unsigned *hours, unsigned *minutes,
 	unsigned *seconds, unsigned *ms)
 {
 	// LLONG milli = (LLONG) ((ccblock*1000)/29.97);
-	*ms = (unsigned)(milli % 1000); // milliseconds
+	*ms = (unsigned) milli % 1000; // milliseconds
```
Comment (cfsmp3) [inline src/lib_ccx/ccx_common_common.c]: What's wrong with the parenthesis? They don't make the code slower, and readability is always better :-)

File: src/lib_ccx/ccx_encoders_helpers.c [hunk 3]
```diff
@@ -235,49 +235,50 @@ unsigned char *close_tag(struct encoder_ctx *ctx, unsigned char *buffer, char *t
 
 unsigned get_decoder_line_encoded(struct encoder_ctx *ctx, unsigned char *buffer, int line_num, struct eia608_screen *data)
 {
-	int col = COL_WHITE;
+	int colour = COL_WHITE;
```
Comment (cfsmp3) [inline src/lib_ccx/ccx_encoders_helpers.c]: Small thing but since this is new... we have "color" in some places. Let's stick to one spelling. Since CC means "closed captions" which is an American term let's use the American spelling, i.e. "color". 
Not saying it's the correct one but it was in the code first.

File: src/lib_ccx/ccx_encoders_scc.c [hunk 4]
```diff
@@ -0,0 +1,371 @@
+// Reference: https://www.govinfo.gov/app/details/CFR-2011-title47-vol1/CFR-2011-title47-vol1-sec15-119/summary
+// Implementatin issue: https://github.com/CCExtractor/ccextractor/issues/1120
+
+#include "lib_ccx.h"
+#include <stdbool.h>
+
+// Adds a parity bit if needed
+unsigned char odd_parity(const unsigned char byte)
+{
+	return byte | !(__builtin_parity(byte) % 2) << 7;
```
Comment (cfsmp3) [inline src/lib_ccx/ccx_encoders_scc.c]: This is GNU. Have you tried to build it with Visual Studio as well?

File: src/lib_ccx/utility.c [hunk 5]
```diff
@@ -506,36 +506,40 @@ char *get_basename(char *filename)
...
-char *get_file_extension(enum ccx_output_format write_format)
+char *get_file_extension(const enum ccx_output_format write_format)
 {
 	switch (write_format)
 	{
+		case CCX_OF_CCD:
+			return ".ccd";
 		case CCX_OF_RAW:
-			return strdup(".raw");
+			return ".raw";
```
Comment (cfsmp3) [inline src/lib_ccx/utility.c]: You are returning a non-const pointer to data in code. Probably really bad things will happen if the caller tries for example to change the file extension in any way (such as make it uppercase).

https://stackoverflow.com/questions/4836534/returning-a-pointer-to-a-literal-or-constant-character-array-string

File: src/lib_ccx/ccx_encoders_scc.c (hunk 4 above)
Comment (NilsIrl) [reply src/lib_ccx/ccx_encoders_scc.c]: Is there a function I should use instead or should I just add a function to replace this one?

File: src/lib_ccx/utility.c (hunk 5 above)
Comment (NilsIrl) [reply src/lib_ccx/utility.c]: I'm surprised I haven't done it. Well anyway fixed it.

File: src/lib_ccx/ccx_common_common.c (hunk 2 above)
Comment (NilsIrl) [reply src/lib_ccx/ccx_common_common.c]: Reverted, but IMO they actually decrease readability as you expect them to do something when they in fact don't do anything

File: src/lib_ccx/ccx_encoders_scc.c (hunk 4 above)
Comment (NilsIrl) [reply src/lib_ccx/ccx_encoders_scc.c]: I haven't tried with Visual Studio though I don't expect it to work as I added it knowing it was an extension

File: src/lib_ccx/ccx_encoders_helpers.c (hunk 3 above)
Comment (NilsIrl) [reply src/lib_ccx/ccx_encoders_helpers.c]: Yeah when I put that in I hadn't seen any other code to follow

File: src/lib_ccx/ccx_encoders_scc.c (hunk 4 above)
Comment (cfsmp3) [reply src/lib_ccx/ccx_encoders_scc.c]: Take a look at 
int cc608_parity(unsigned int byte)

File: src/lib_ccx/ccx_encoders_scc.c [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+// than preamble codes as preamble codes can only be precise by 4 columns
+// e.g.
+// indent * 4 + tab_offset
+// where tab_offset can be 1 to 3. If no offset wants to be used tab_offset disappears
+
+const unsigned char TO_CHANNEL_1 = 0x17;
+const unsigned char TO_CHANNEL_2 = 0x1f;
+const unsigned char TO_OFFSET = 0x20;
+
+// The second byte for tab offsets, is the 0x20 + column
```
Comment (NilsIrl) [inline src/lib_ccx/ccx_encoders_scc.c]: Move this comment up to where it's relevant

File: src/lib_ccx/ccx_common_constants.h [hunk 7]
```diff
@@ -181,7 +181,9 @@ enum ccx_output_format
 	CCX_OF_G608       = 11,
 	CCX_OF_CURL       = 12,
 	CCX_OF_SSA        = 13,
-    CCX_OF_MCC        = 14
+    CCX_OF_MCC        = 14,
```
Comment (NilsIrl) [inline src/lib_ccx/ccx_common_constants.h]: re fix formatting

File: src/lib_ccx/ccx_encoders_srt.c (hunk 1 above)
Comment (NilsIrl) [reply src/lib_ccx/ccx_encoders_srt.c]: Reverted

File: src/lib_ccx/ccx_encoders_scc.c (hunk 4 above)
Comment (NilsIrl) [reply src/lib_ccx/ccx_encoders_scc.c]: Fixed

**Comment by NilsIrl** [comment]
For some reason the `.srt` output doesn't work anymore

**Comment by cfsmp3** [comment]
Because?

**Comment by NilsIrl** [comment]
> Because?

As I said "some reason".

I have no clue. I thought for some time it had to do with using enums instead of ints but that's not the problem. Also webvtt isn't affected so....

**Comment by NilsIrl** [comment]
Okay, fixed it

I didn't notice the `if` statement after that which meant these values still had to be initialized

**Comment by NilsIrl** [comment]
Here is a list of post merge operations to perform:
* Unify the subtitle modifying features (Capitalization + Bad word filtration)
* Unify dealing with OCR stuff
* Unifying addition of delay which will fix #1103

**Comment by cfsmp3** [comment]
Still fails:

```
1>ccx_encoders_scc.obj : error LNK2019: unresolved external symbol _dprintf referenced in function _add_padding
1>Debug\ccextractorwin.exe : fatal error LNK1120: 1 unresolved externals
1>Done building project "ccextractor.vcxproj" -- FAILED.
```

Just get rid of that dprintf() :-)

**Comment by NilsIrl** [comment]
> Just get rid of that dprintf() :-)

Yeah I realised MSVC actually has a dprintf but it's a totally different function

**Comment by cfsmp3** [comment]
> > Just get rid of that dprintf() :-)
> 
> Yeah I realised MSVC actually has a dprintf but it's a totally different function

In general (not just for us) - submitting untested stuff decreases your street cred :-) It's OK we find bugs when we test corner cases, but a PR that doesn't even compile is not good...

We want CCExtractor to be portable (as is it builds for anything that has a C compiler), so in general, any function that is an extension of a specific compiler (even if it is GCC) is best not to use.

**Comment by NilsIrl** [comment]
I changed the architecture to make it less "hacky". That is, this new architecture, is more linear and doesn't try to make sense of what is seemingly random. Codes are now hardcoded. This will make it easier to add features (color) without having to make it more complex.

**Comment by NilsIrl** [comment]
I'll also add that this last commit gives the exact same output as the previous one. (except that it doesn't produce trailing space anymore)

**Comment by cfsmp3** [comment]
@NilsIrl can you resolve conflicts? (and let us know the current status)

**Comment by NilsIrl** [comment]
* Color support needs to be added
* And line endings ("\n" vs "\r\n") needs to be implemented (for the moment "\n" is used everywhere

**Comment by cfsmp3** [comment]
Go ahead. About line endings though keep in mind that it can be format
specific to, so check the specs. If you they are say it's always one thing
then we write that thing instead of making it platform dependent

**Comment by cfsmp3** [comment]
@NilsIrl Note that there's conflicts again.

**Comment by NilsIrl** [comment]
@cfsmp3 there is `fdprintf` in `ccx_common_common` that is a portable implementation of `dprintf`.

**Comment by cfsmp3** [comment]
I wouldn't worry so much about it being feature-complete from day 1 - it's more important that it doesn't break anything that is already there and that for some cases it works perfectly (which is better than working so-so for all cases).

**Comment by NilsIrl** [comment]
Another thing that isn't implemented is preamble codes that set the font and the placement on the screen at the same time.
